Add solution folder argument to dotnet-sln add

LordMike opened the pull request:
I've begun work on the first half of what I think "dotnet-sln add" should be allowed to do:

- Target arbitrary solution folders within an SLN
- Add arbitrary files as solution-level files, and not just projects

If possible, I'd like some feedback before proceeding to add the ability for "dotnet-sln add" to add arbitrary files. So, I suppose this is kind of a WIP which isn't exactly preferable... Sorry about that.

Is there a test project for "dotnet sln" ?

Related issues:
- Issue https://github.com/dotnet/cli/issues/8678, which discuses a "add" feature that allows arbitrary files to be added; where I outline my intended commandline layout
- PR https://github.com/dotnet/cli/pull/10873, which touches on a similar subject, but also adds a unit test project. Naming between that PR, and this one, should be coordinated.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/cli?pullRequest=11000) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/cli?pullRequest=11000) before we can accept your contribution.<br/><br/>:x: LordMike [sign now](https://cla.dotnetfoundation.org/dotnet/cli?pullRequest=11000)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/cli?pullRequest=11000) it.</sub>

peterhuene commented on src/dotnet/commands/dotnet-sln/LocalizableStrings.resx:
```diff
@@ -162,4 +162,7 @@
     <value>Project(s)</value>
   </data>
+  <data name="AddProjectSolutionFolderArgumentDescription" xml:space="preserve">
+    <value>The destination solution folder to add the items to</value>
```
```suggestion
    <value>The destination solution folder path to add the project(s) to.</value>
```

peterhuene replied:
Ah, I just saw the issue you linked to regarding the future use of adding solution items via `dotnet sln add`.  I think we should go with "project" rather than "item" for now; we can always change this if we support arbitrary solution items via this command in the future.

peterhuene commented on src/dotnet/SlnFileExtensions.cs:
```diff
@@ -70,6 +70,13 @@ public static void AddProject(this SlnFile slnFile, string fullProjectPath)
 
+                // Identify the intended solution folders
+                IList<string> solutionFolders;
+                if (!string.IsNullOrEmpty(relativeRoot))
```
We generally prefer always using curly braces around if-else statements, both as a style and that it's easier to maintain.

peterhuene commented on src/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -51,9 +51,13 @@ public override int Execute()
             var preAddProjectCount = slnFile.Projects.Count;
 
+            string relativeRoot = null;
+            if (_appliedCommand.AppliedOptions != null && _appliedCommand.AppliedOptions.Contains("-s"))
```
Nit: we prefer checking against the long-form option so when someone reads this they don't have to go check the parser to remember what option we're checking for.

peterhuene replied:
See previous comment regarding curly braces too.

LordMike commented:
Thanks @peterhuene .. Should I work on adding arbitrary items in the same PR, or split that work into this and another PR?

peterhuene commented:
Let's do another PR for now.  Adding arbitrary items to the solution instead of just projects hasn't been considered fully by our team and I'd like to get feedback from @KathleenDollard who is our PM.

LordMike commented:
@peterhuene are you certain #10873 will be merged, and that it is the one I should use? 

I can get to work on unit tests, by merging that PR's branch into mine, thereby getting the changes I need. Would this be ok? (it would break the ability to squash the merge in the future)

peterhuene commented:
@LordMike I'm fairly confident we'll merge #10873 after we land on the exact name we want to use for the option.  I would create a new branch from your feature branch and rebase that new branch off of theirs if you want to get started on some unit tests.

However, there's no rush, so feel free to wait until the other PR merges, which will hopefully be soon.  We still have plenty of time to get this in the 3.0 release.

peterhuene commented:
@LordMike #10873 has now been merged, so you'll have some conflicts to resolve if you rebase your feature branch off of master.

There should now be a `sln add` tests project where you can add tests for this feature.

Please let me know if you have any questions.  Thanks!

LordMike commented:
@peterhuene it works, but it's not entirely pretty (there are now contradicting options in the `SlnFileExtensions` helper utility. F.ex., you can pass both a boolean to say whether or not a project should be in the root, and also a path where it should be. 

Here I'd rather go with a single parameter that simultaneously determines if a project goes to the root (null/empty string), or in a folder (path to the folder)..

Is it my task or Microsofts, to ensure code style / quality in this manner?

peterhuene commented:
Having two mutually exclusive options can be ok, especially when it makes the intent of the options clear.  I personally don't like the convention of "empty" arguments having semantic meaning, and I'm not sure I would favor something like `--solution-folder /` or `--solution-folder .` to mean "don't create any solution folders".

But I'll take a step back and think about this.  If we do keep both options, I would expect specifying both to be an error.  Specifying neither would keep the default behavior of creating it based on the relative path to the project from the solution.

@KathleenDollard do you have any thoughts on this?

LordMike commented:
@peterhuene you misunderstood me. I have no qualms with the CLI options.. I'm talking about internally in the code.. The SlnFileExtensions now has this method:

https://github.com/dotnet/cli/blob/815803f45edd2944a0816315fd6c98a1236120a4/src/dotnet/SlnFileExtensions.cs#L19

In it, there are two arguments: `bool inRoot, string relativeRoot = null` -- and I was asking if those should be merged somehow. :)

peterhuene commented:
Oohh, I see.  Yes, that should probably be changed such that some of this defaulting logic is moved to the caller of `AddProject`, probably removing `inRoot` in favor of a null `relativeRoot` implying "add to root" rather than calculating a default solution folder path.

That said, what approach do you think we should take for merging these confusing parameters to `AddProject`?

LordMike commented:
I would probably do just that. Move the logic of determining solution folders out of AddProject (it might be used for AddItem in the future ..), and then reduce AddProject to a single string/list of strings that determine the solution folder.

LordMike commented:
@peterhuene I've updated the failing test (the help text one), fixed issues and refactored the code. The build has one failing test in `dotnet run` .. but I'm unsure what to do there.

peterhuene commented on src/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -32,7 +34,32 @@ internal class AddProjectToSolutionCommand : CommandBase
-            _inRoot = appliedCommand.ValueOrDefault<bool>(InRootOption);
+
+            var inRoot = appliedCommand.ValueOrDefault<bool>(InRootOption);
+            var relativeRoot = _appliedCommand.ValueOrDefault<string>(SolutionFolderOption);
```
I think we should error (i.e. throw a `GracefulException`) if both options are specified since they are mutually exclusive.

peterhuene commented on src/dotnet/commands/dotnet-sln/LocalizableStrings.resx:
```diff
@@ -165,4 +165,7 @@
     <value>Place project in root of the solution, rather than creating a solution folder.</value>
   </data>
+  <data name="AddProjectSolutionFolderArgumentDescription" xml:space="preserve">
+    <value>The destination solution folder path to add the projects to</value>
```
```suggestion
    <value>The destination solution folder path to add the projects to.</value>
```

LordMike replied:
I've noticed the `xlf` auto-updated files aren't updated alongside the original `resx` file.. should that be done?

LordMike replied:
Nevermind. I just found out.

peterhuene commented on src/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -16,12 +17,13 @@ internal class AddProjectToSolutionCommand : CommandBase
         private readonly string _fileOrDirectory;
-        private readonly bool _inRoot;
-        
+        private readonly Func<SlnFile, string, IList<string>> _determineSolutionFolder;
```
Rather than storing a `Func`, could we store the `IList<string>` of the solution folders?  I think it would be a little less confusing.

LordMike replied:
It might make a mess of the code.[ In one of the three cases](https://github.com/dotnet/cli/blob/e84a3b763618d5c15645d76d574f835f4c5758a7/src/dotnet/commands/dotnet-sln/add/Program.cs#L55-L61), the solution folders list depends on the concrete project. The Func / delegate approach was the most concise approach I could come up with.

Should I move the if-logic to the loop over the projects?

peterhuene replied:
Personally, I think it would be easier to reason with a method that does the if-logic rather than the `Func`.

If we were storing `_inRoot` and `_relativeRoot` and moved the extension method logic into this class, then the method would simply be something like this:

```c#
private IList<string> GetSolutionFolders(SlnFile slnFile, string fullProjectPath)
{
    if (_inRoot)
    {
        return null;
    }

    if (!string.IsNullOrEmpty(_relativeRoot))
    {
        return _relativeRoot.Split(Path.DirectorySeparatorChar);
    }

    return GetSolutionFoldersFromProjectPath(
        Path.GetRelativePath(
            PathUtility.EnsureTrailingSlash(slnFile.BaseDirectory),
            fullProjectPath));
}
```

peterhuene replied:
@LordMike any thoughts to this remaining comment?  I see you moved the previous extension method into this class, so the above should be usable as is.

LordMike replied:
Woops. Hadn't seen the update here.

I could do that, but would you want to recalculate the root in the `_relativeRoot` case, for each project to add?.. Seems wasteful, even though it's likely a matter of at most 1-3 times in a process lifetime (dotnet sln add ..)

peterhuene replied:
Sure, we can split it just once if it was specified and store that as a field instead of the string.

peterhuene commented:
@LordMike the test failure you were seeing should be fixed in master; I think the base for your feature branch might not have included it yet.  It's a transient timing failure, so if we see it again we can just rerun the test.

LordMike commented:
@peterhuene could you retrigger the failing test?

peterhuene commented:
Re-running the test now.  I'll review your recent pushes soon and I think we'll be good to go!  Thanks!

peterhuene commented on src/dotnet/SlnProjectExtensions.cs:
```diff
@@ -1,20 +1,17 @@
     internal static class SlnProjectExtensions
     {
-        public static IList<string> GetSolutionFoldersFromProject(this SlnProject project)
+        public static IList<string> GetSolutionFoldersFromProjectPath(string projectFilePath)
```
I just noticed this change no longer makes this an extension method.

Given that we only have one caller of this in `AddProjectToSolutionCommand `, perhaps we should just move these few methods into that class and delete this file?

LordMike replied:
I was unsure if I should do this, but since you think the same .. :)

peterhuene commented on src/dotnet/commands/dotnet-sln/LocalizableStrings.resx:
```diff
@@ -165,4 +165,10 @@
+    <value>The destination solution folder path to add the projects to.</value>
+  </data>
+  <data name="SolutionFolderAndInRootMutuallyExclusive" xml:space="preserve">
+    <value>The --solution-folder and --in-root options are mutually exclusive</value>
```
I think we can probably use a slightly friendlier error message that doesn't use jargon like "mutually exclusive" which might confuse people unfamiliar with the term.

Perhaps this:

```suggestion
    <value>The --solution-folder and --in-root options cannot be used together; use only one of the options.</value>
```

What do you think?

LordMike replied:
Sounds fine with me. I'm indifferent to texts, as long as the point gets across :)

peterhuene commented:
Hi @LordMike.  I left just a few more comments.  I think we're almost there and it's looking really good!

I really appreciate your patience and for being so open to my feedback.

LordMike commented:
I've made the two changes (text, refactoring) and I think I got it all (updated XLF files and updated the unit test for mutual exclusivity).. That should cover it. :)

peterhuene commented:
@KathleenDollard for your information, here's what the UX will be for this new option:

```
dotnet sln add --solution-folder 'folder1\folder2' myproject.csproj
```

KathleenDollard commented:
Looks good! Drop a line to @mairaw and I when it is merged so we can get a plan for docs, release notes, etc.

peterhuene commented on src/dotnet/commands/dotnet-sln/add/Program.cs:
```diff
@@ -17,11 +19,14 @@ internal class AddProjectToSolutionCommand : CommandBase
         private readonly string _fileOrDirectory;
         private readonly bool _inRoot;
-        
+        private readonly string _relativeRoot;
```
Perhaps we should just store `_relativeRootSolutionFolders`?

That is to say, in the constructor, using a local `relativeRoot` instead of the field:

```
 _relativeRootSolutionFolders = relativeRoot?.Split(Path.DirectorySeparatorChar);
```

And thus in `DetermineSolutionFolder`:

```
...
if (_relativeRootSolutionFolders != null)
{
    return _relativeRootSolutionFolders;
}
...
```

I think that would be a little bit cleaner than having `DetermineSolutionFolder` have side-effects (i.e. assignment to `_relativeRootSolutionFolders `).

What do you think?

LordMike replied:
That _would_ make it a little cleaner. Is this the last change? :)

peterhuene replied:
I think so 😄 🤞 🎉

nguerrera commented:
Thanks, @LordMike

peterhuene commented:
Thank you very much for the contribution, @LordMike!  And for being an excellent collaborator to get this feature landed (I know the review was lengthy 😄).

Look for the feature in Preview 5 of the .NET Core 3.0 SDK.

LightVolk commented:
@LordMike , if possible, can this fucntionality be added: list of solution folders.
Use case: 
1. list of solution folders
2. add new project under on of this solution folder.

LordMike commented:
@LightVolk I'm not in charge of that.. but I suppose you should add a new feature request for it.. ?
